File: poc/atproto/dds-reference/packages/client-ts/src/agent.ts
import { Agent } from "@atproto/api";
import type {
  OrgDdsWgV1Algorithm,
  OrgDdsWgV1Classification,
  OrgDdsWgV1Decision,
  OrgDdsWgV1Group,
  OrgDdsWgV1Metric,
  OrgDdsWgV1Project,
  OrgDdsWgV1Proposal,
  OrgDdsWgV1Response,
  OrgDdsWgV1Statement,
  OrgDdsWgV1Summary,
  OrgDdsWgV1Vote,
} from "./lexicon/index.js";
import type * as Defs from "./lexicon/types/org/dds-wg/v1/defs.js";
import type * as StrongRef from "./lexicon/types/com/atproto/repo/strongRef.js";
import { resolveDidPds } from "./identity.js";
import { COLLECTIONS, LEXICON_VERSION } from "./version.js";

const DEFAULT_PDS_FALLBACK = "https://bsky.social";

/**
 * Anything that can be turned into an @atproto/api Agent: an Agent itself,
 * a `service` URL string, or an OAuth session object that the Agent constructor accepts.
 */
export type AgentLike = Agent | ConstructorParameters<typeof Agent>[0];

export interface CreateRecordResult {
  uri: string;
  cid: string;
}

/**
 * Wraps an authenticated @atproto/api Agent with typed convenience methods for the
 * DDS lexicon. All write methods set `createdAt` to now and `schemaVersion` to the
 * current lexicon version unless overridden.
 *
 * Mutations are scoped to the authenticated repo (the session's DID). Reads either
 * accept an explicit `repo` or default to the session's DID. When reading from another
 * user's repo, the client resolves their DID to their PDS endpoint and dispatches the
 * read there (the authenticated session can only see records on its own PDS).
 */
export class DDSClient {
  readonly agent: Agent;
  /** Per-PDS unauthenticated agents used for cross-repo reads. */
  private readonly readAgents = new Map<string, Agent>();

  constructor(input: AgentLike) {
    this.agent = input instanceof Agent ? input : new Agent(input);
  }

  /** Get (or build) an unauthenticated agent pointed at the PDS hosting `did`. */
  private async readAgentFor(did: string): Promise<Agent> {
    if (this.did && did === this.did) return this.agent;
    const cached = this.readAgents.get(did);
    if (cached) return cached;
    const pds = (await resolveDidPds(did)) ?? DEFAULT_PDS_FALLBACK;
    const agent = new Agent(pds);
    this.readAgents.set(did, agent);
    return agent;
  }

  /** The DID of the authenticated session, if any. */
  get did(): string | undefined {
    return this.agent.did;
  }

  private requireDid(): string {
    const did = this.did;
    if (!did) {
      throw new Error("DDSClient: no authenticated session (agent.did is undefined)");
    }
    return did;
  }

  private now(): string {
    return new Date().toISOString();
  }

  private userAgentSelf(): Defs.UserAgent & { $type: "org.dds-wg.v1.defs#userAgent" } {
    return { $type: "org.dds-wg.v1.defs#userAgent", did: this.requireDid() };
  }

  private async createRecord(
    collection: string,
    record: Record<string, unknown>,
  ): Promise<CreateRecordResult> {
    const repo = this.requireDid();
    const res = await this.agent.com.atproto.repo.createRecord({
      repo,
      collection,
      record: { schemaVersion: LEXICON_VERSION, createdAt: this.now(), ...record },
    });
    return { uri: res.data.uri, cid: res.data.cid };
  }

  // ----- Records -----

  createStatement(input: {
    text: string;
    replyTo?: StrongRef.Main;
    step?: Defs.ProjectStepRef;
  }): Promise<CreateRecordResult> {
    const record: OrgDdsWgV1Statement.Main = {
      $type: COLLECTIONS.statement,
      text: input.text,
      author: this.userAgentSelf(),
      replyTo: input.replyTo,
      step: input.step,
      createdAt: this.now(),
    };
    return this.createRecord(COLLECTIONS.statement, record as unknown as Record<string, unknown>);
  }

  createOrdinalVote(input: {
    target: StrongRef.Main;
    value: -1 | 0 | 1;
    step?: Defs.ProjectStepRef;
  }): Promise<CreateRecordResult> {
    const record: OrgDdsWgV1Vote.Main = {
      $type: COLLECTIONS.vote,
      author: this.userAgentSelf(),
      target: input.target,
      value: { $type: "org.dds-wg.v1.vote#ordinal", value: input.value },
      step: input.step,
      createdAt: this.now(),
    };
    return this.createRecord(COLLECTIONS.vote, record as unknown as Record<string, unknown>);
  }

  createProposal(input: {
    name: string;
    description?: string;
    step?: Defs.ProjectStepRef;
  }): Promise<CreateRecordResult> {
    const record: OrgDdsWgV1Proposal.Main = {
      $type: COLLECTIONS.proposal,
      name: input.name,
      description: input.description,
      creator: this.userAgentSelf(),
      step: input.step,
      createdAt: this.now(),
    };
    return this.createRecord(COLLECTIONS.proposal, record as unknown as Record<string, unknown>);
  }

  createDecision(input: {
    name: string;
    summary: string;
    acceptedProposals?: StrongRef.Main[];
    step?: Defs.ProjectStepRef;
  }): Promise<CreateRecordResult> {
    const record: OrgDdsWgV1Decision.Main = {
      $type: COLLECTIONS.decision,
      name: input.name,
      summary: input.summary,
      creator: this.userAgentSelf(),
      acceptedProposals: input.acceptedProposals,
      step: input.step,
      createdAt: this.now(),
    };
    return this.createRecord(COLLECTIONS.decision, record as unknown as Record<string, unknown>);
  }

  createSummary(input: {
    name: string;
    description: string;
    subject: StrongRef.Main;
    step?: Defs.ProjectStepRef;
  }): Promise<CreateRecordResult> {
    const record: OrgDdsWgV1Summary.Main = {
      $type: COLLECTIONS.summary,
      name: input.name,
      description: input.description,
      creator: this.userAgentSelf(),
      subject: input.subject,
      step: input.step,
      createdAt: this.now(),
    };
    return this.createRecord(COLLECTIONS.summary, record as unknown as Record<string, unknown>);
  }

  createResponse(input: {
    text: string;
    target: StrongRef.Main;
  }): Promise<CreateRecordResult> {
    const record: OrgDdsWgV1Response.Main = {
      $type: COLLECTIONS.response,
      text: input.text,
      target: input.target,
      author: this.userAgentSelf(),
      createdAt: this.now(),
    };
    return this.createRecord(COLLECTIONS.response, record as unknown as Record<string, unknown>);
  }

  createGroup(input: {
    name?: string;
    description?: string;
    members: OrgDdsWgV1Group.Main["members"];
    step?: Defs.ProjectStepRef;
  }): Promise<CreateRecordResult> {
    const record: OrgDdsWgV1Group.Main = {
      $type: COLLECTIONS.group,
      name: input.name,
      description: input.description,
      creator: this.userAgentSelf(),
      members: input.members,
      step: input.step,
      createdAt: this.now(),
    };
    return this.createRecord(COLLECTIONS.group, record as unknown as Record<string, unknown>);
  }

  createMetric(input: {
    name: string;
    metricType: string;
    subject: StrongRef.Main;
    value: OrgDdsWgV1Metric.Main["value"];
    step?: Defs.ProjectStepRef;
  }): Promise<CreateRecordResult> {
    const record: OrgDdsWgV1Metric.Main = {
      $type: COLLECTIONS.metric,
      name: input.name,
      creator: this.userAgentSelf(),
      metricType: input.metricType,
      value: input.value,
      subject: input.subject,
      step: input.step,
      createdAt: this.now(),
    };
    return this.createRecord(COLLECTIONS.metric, record as unknown as Record<string, unknown>);
  }

  createClassification(input: {
    subject: StrongRef.Main;
    label: string;
    /** 0-10000 fixed-point, where 10000 = 100% confidence */
    confidence?: number;
  }): Promise<CreateRecordResult> {
    const record: OrgDdsWgV1Classification.Main = {
      $type: COLLECTIONS.classification,
      creator: this.userAgentSelf(),
      subject: input.subject,
      label: input.label,
      confidence: input.confidence,
      createdAt: this.now(),
    };
    return this.createRecord(
      COLLECTIONS.classification,
      record as unknown as Record<string, unknown>,
    );
  }

  createAlgorithm(input: {
    name: string;
    description?: string;
    metadata?: unknown;
  }): Promise<CreateRecordResult> {
    const record: OrgDdsWgV1Algorithm.Main = {
      $type: COLLECTIONS.algorithm,
      name: input.name,
      description: input.description,
      creator: this.requireDid(),
      metadata: input.metadata !== undefined ? JSON.stringify(input.metadata) : undefined,
      createdAt: this.now(),
    };
    return this.createRecord(COLLECTIONS.algorithm, record as unknown as Record<string, unknown>);
  }

  createProject(input: {
    name: string;
    description?: string;
    phases: OrgDdsWgV1Project.Phase[];
  }): Promise<CreateRecordResult> {
    const record: OrgDdsWgV1Project.Main = {
      $type: COLLECTIONS.project,
      name: input.name,
      description: input.description,
      creator: this.userAgentSelf(),
      phases: input.phases,
      createdAt: this.now(),
    };
    return this.createRecord(COLLECTIONS.project, record as unknown as Record<string, unknown>);
  }

  // ----- Queries -----

  /** List records in a DDS collection for a given repo (defaults to the session's DID). */
  async listRecords(input: {
    collection: string;
    repo?: string;
    limit?: number;
    cursor?: string;
  }): Promise<{ cursor?: string; records: Array<{ uri: string; cid: string; value: unknown }> }> {
    const repo = input.repo ?? this.requireDid();
    const agent = await this.readAgentFor(repo);
    const res = await agent.com.atproto.repo.listRecords({
      repo,
      collection: input.collection,
      limit: input.limit,
      cursor: input.cursor,
    });
    return { cursor: res.data.cursor, records: res.data.records };
  }

  /** Fetch a single record by AT-URI. The repo and collection are parsed from the URI. */
  async getRecord(uri: string): Promise<{ uri: string; cid: string; value: unknown }> {
    const parsed = parseAtUri(uri);
    const agent = await this.readAgentFor(parsed.repo);
    const res = await agent.com.atproto.repo.getRecord({
      repo: parsed.repo,
      collection: parsed.collection,
      rkey: parsed.rkey,
    });
    return { uri: res.data.uri, cid: res.data.cid ?? "", value: res.data.value };
  }

  /** List projects in a repo (defaults to the session's DID). */
  listProjects(input?: { repo?: string; limit?: number; cursor?: string }) {
    return this.listRecords({ ...input, collection: COLLECTIONS.project });
  }

  /** List statements in a repo. */
  listStatements(input?: { repo?: string; limit?: number; cursor?: string }) {
    return this.listRecords({ ...input, collection: COLLECTIONS.statement });
  }

  /** List votes in a repo. */
  listVotes(input?: { repo?: string; limit?: number; cursor?: string }) {
    return this.listRecords({ ...input, collection: COLLECTIONS.vote });
  }
}

interface ParsedAtUri {
  repo: string;
  collection: string;
  rkey: string;
}

export function parseAtUri(uri: string): ParsedAtUri {
  if (!uri.startsWith("at://")) {
    throw new Error(`Not an at:// URI: ${uri}`);
  }
  const [repo, collection, rkey, ...rest] = uri.slice(5).split("/");
  if (!repo || !collection || !rkey || rest.length > 0) {
    throw new Error(`Malformed AT-URI: ${uri}`);
  }
  return { repo, collection, rkey };
}

File: poc/atproto/dds-reference/python/dds_client/dds_client/agent.py
"""DDSClient: high-level wrapper around an authenticated atproto.Client.

Mirrors `packages/client-ts/src/agent.ts`.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Optional

from atproto import Client
from atproto_client.models import dot_dict

from .version import COLLECTIONS, LEXICON_VERSION


@dataclass(frozen=True)
class CreateRecordResult:
    uri: str
    cid: str


@dataclass(frozen=True)
class ParsedAtUri:
    repo: str
    collection: str
    rkey: str


def parse_at_uri(uri: str) -> ParsedAtUri:
    if not uri.startswith("at://"):
        raise ValueError(f"Not an at:// URI: {uri}")
    parts = uri[len("at://"):].split("/")
    if len(parts) != 3 or not all(parts):
        raise ValueError(f"Malformed AT-URI: {uri}")
    return ParsedAtUri(repo=parts[0], collection=parts[1], rkey=parts[2])


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


class DDSClient:
    """High-level, typed wrapper around an authenticated `atproto.Client`.

    Mutations are scoped to the authenticated repo (the session's DID). Reads
    accept an explicit `repo` or default to the session's DID.
    """

    def __init__(self, client: Client):
        self.client = client

    @property
    def did(self) -> Optional[str]:
        return getattr(self.client.me, "did", None) if self.client.me else None

    def _require_did(self) -> str:
        d = self.did
        if d is None:
            raise RuntimeError("DDSClient: no authenticated session")
        return d

    def _user_agent_self(self) -> dict[str, Any]:
        return {"$type": "org.dds-wg.v1.defs#userAgent", "did": self._require_did()}

    def _create_record(self, collection: str, record: dict[str, Any]) -> CreateRecordResult:
        repo = self._require_did()
        full = {"schemaVersion": LEXICON_VERSION, "createdAt": _now_iso(), **record}
        # atproto.Client wraps com.atproto.repo.createRecord
        resp = self.client.com.atproto.repo.create_record(
            data={"repo": repo, "collection": collection, "record": dot_dict.DotDict(full)},
        )
        return CreateRecordResult(uri=resp.uri, cid=resp.cid)

    # ---------------- record creation ----------------

    def create_statement(
        self,
        *,
        text: str,
        reply_to: Optional[dict[str, Any]] = None,
        step: Optional[dict[str, Any]] = None,
    ) -> CreateRecordResult:
        record = {
            "$type": COLLECTIONS.statement,
            "text": text,
            "author": self._user_agent_self(),
            **({"replyTo": reply_to} if reply_to else {}),
            **({"step": step} if step else {}),
        }
        return self._create_record(COLLECTIONS.statement, record)

    def create_ordinal_vote(
        self,
        *,
        target: dict[str, Any],
        value: int,
        step: Optional[dict[str, Any]] = None,
    ) -> CreateRecordResult:
        if value not in (-1, 0, 1):
            raise ValueError("ordinal vote value must be -1, 0, or 1")
        record = {
            "$type": COLLECTIONS.vote,
            "author": self._user_agent_self(),
            "target": target,
            "value": {"$type": "org.dds-wg.v1.vote#ordinal", "value": value},
            **({"step": step} if step else {}),
        }
        return self._create_record(COLLECTIONS.vote, record)

    def create_continuous_vote(
        self,
        *,
        target: dict[str, Any],
        value: float,
        step: Optional[dict[str, Any]] = None,
    ) -> CreateRecordResult:
        if not -1.0 <= value <= 1.0:
            raise ValueError("continuous vote value must be in [-1, 1]")
        fixed = round(value * 10000)
        record = {
            "$type": COLLECTIONS.vote,
            "author": self._user_agent_self(),
            "target": target,
            "value": {"$type": "org.dds-wg.v1.vote#continuous", "value": fixed},
            **({"step": step} if step else {}),
        }
        return self._create_record(COLLECTIONS.vote, record)

    def create_proposal(
        self,
        *,
        name: str,
        description: Optional[str] = None,
        step: Optional[dict[str, Any]] = None,
    ) -> CreateRecordResult:
        record = {
            "$type": COLLECTIONS.proposal,
            "name": name,
            "creator": self._user_agent_self(),
            **({"description": description} if description else {}),
            **({"step": step} if step else {}),
        }
        return self._create_record(COLLECTIONS.proposal, record)

    def create_decision(
        self,
        *,
        name: str,
        summary: str,
        accepted_proposals: Optional[list[dict[str, Any]]] = None,
        step: Optional[dict[str, Any]] = None,
    ) -> CreateRecordResult:
        record = {
            "$type": COLLECTIONS.decision,
            "name": name,
            "summary": summary,
            "creator": self._user_agent_self(),
            **({"acceptedProposals": accepted_proposals} if accepted_proposals else {}),
            **({"step": step} if step else {}),
        }
        return self._create_record(COLLECTIONS.decision, record)

    def create_summary(
        self,
        *,
        name: str,
        description: str,
        subject: dict[str, Any],
        step: Optional[dict[str, Any]] = None,
    ) -> CreateRecordResult:
        record = {
            "$type": COLLECTIONS.summary,
            "name": name,
            "description": description,
            "creator": self._user_agent_self(),
            "subject": subject,
            **({"step": step} if step else {}),
        }
        return self._create_record(COLLECTIONS.summary, record)

    def create_response(self, *, text: str, target: dict[str, Any]) -> CreateRecordResult:
        record = {
            "$type": COLLECTIONS.response,
            "text": text,
            "target": target,
            "author": self._user_agent_self(),
        }
        return self._create_record(COLLECTIONS.response, record)

    def create_group(
        self,
        *,
        members: list[dict[str, Any]],
        name: Optional[str] = None,
        description: Optional[str] = None,
        step: Optional[dict[str, Any]] = None,
    ) -> CreateRecordResult:
        record = {
            "$type": COLLECTIONS.group,
            "creator": self._user_agent_self(),
            "members": members,
            **({"name": name} if name else {}),
            **({"description": description} if description else {}),
            **({"step": step} if step else {}),
        }
        return self._create_record(COLLECTIONS.group, record)

    def create_metric(
        self,
        *,
        name: str,
        metric_type: str,
        subject: dict[str, Any],
        value: dict[str, Any],
        step: Optional[dict[str, Any]] = None,
    ) -> CreateRecordResult:
        record = {
            "$type": COLLECTIONS.metric,
            "name": name,
            "creator": self._user_agent_self(),
            "metricType": metric_type,
            "value": value,
            "subject": subject,
            **({"step": step} if step else {}),
        }
        return self._create_record(COLLECTIONS.metric, record)

    # ---------------- algorithm-authored variants ----------------
    #
    # These are used by services like the clustering server, where the `creator` of a record is
    # an algorithm instance rather than the authenticated user. The session is still the user's
    # (since AT Protocol writes are signed by an account), but the in-record `creator` shape is
    # the `algorithmAgent` union variant referencing an `org.dds-wg.v1.algorithm` record.

    @staticmethod
    def _algorithm_agent(algorithm_uri: str, algorithm_cid: str = "") -> dict[str, Any]:
        return {
            "$type": "org.dds-wg.v1.defs#algorithmAgent",
            "algorithm": {"uri": algorithm_uri, "cid": algorithm_cid},
        }

    def create_group_as_algorithm(
        self,
        *,
        algorithm_uri: str,
        members: list[dict[str, Any]],
        name: Optional[str] = None,
        description: Optional[str] = None,
        step: Optional[dict[str, Any]] = None,
        algorithm_cid: str = "",
    ) -> CreateRecordResult:
        record = {
            "$type": COLLECTIONS.group,
            "creator": self._algorithm_agent(algorithm_uri, algorithm_cid),
            "members": members,
            **({"name": name} if name else {}),
            **({"description": description} if description else {}),
            **({"step": step} if step else {}),
        }
        return self._create_record(COLLECTIONS.group, record)

    def create_metric_as_algorithm(
        self,
        *,
        algorithm_uri: str,
        name: str,
        metric_type: str,
        subject: dict[str, Any],
        value: dict[str, Any],
        step: Optional[dict[str, Any]] = None,
        algorithm_cid: str = "",
    ) -> CreateRecordResult:
        record = {
            "$type": COLLECTIONS.metric,
            "name": name,
            "creator": self._algorithm_agent(algorithm_uri, algorithm_cid),
            "metricType": metric_type,
            "value": value,
            "subject": subject,
            **({"step": step} if step else {}),
        }
        return self._create_record(COLLECTIONS.metric, record)

    def create_summary_as_algorithm(
        self,
        *,
        algorithm_uri: str,
        name: str,
        description: str,
        subject: dict[str, Any],
        step: Optional[dict[str, Any]] = None,
        algorithm_cid: str = "",
    ) -> CreateRecordResult:
        record = {
            "$type": COLLECTIONS.summary,
            "name": name,
            "description": description,
            "creator": self._algorithm_agent(algorithm_uri, algorithm_cid),
            "subject": subject,
            **({"step": step} if step else {}),
        }
        return self._create_record(COLLECTIONS.summary, record)

    def create_classification_as_algorithm(
        self,
        *,
        algorithm_uri: str,
        subject: dict[str, Any],
        label: str,
        confidence: Optional[float] = None,
        algorithm_cid: str = "",
    ) -> CreateRecordResult:
        record: dict[str, Any] = {
            "$type": COLLECTIONS.classification,
            "creator": self._algorithm_agent(algorithm_uri, algorithm_cid),
            "subject": subject,
            "label": label,
        }
        if confidence is not None:
            if not 0.0 <= confidence <= 1.0:
                raise ValueError("confidence must be in [0, 1]")
            record["confidence"] = round(confidence * 10000)
        return self._create_record(COLLECTIONS.classification, record)

    # ---------------- end algorithm variants ----------------

    def create_classification(
        self,
        *,
        subject: dict[str, Any],
        label: str,
        confidence: Optional[float] = None,
    ) -> CreateRecordResult:
        record = {
            "$type": COLLECTIONS.classification,
            "creator": self._user_agent_self(),
            "subject": subject,
            "label": label,
        }
        if confidence is not None:
            if not 0.0 <= confidence <= 1.0:
                raise ValueError("confidence must be in [0, 1]")
            record["confidence"] = round(confidence * 10000)
        return self._create_record(COLLECTIONS.classification, record)

    def create_algorithm(
        self,
        *,
        name: str,
        description: Optional[str] = None,
        metadata: Optional[Any] = None,
    ) -> CreateRecordResult:
        import json

        record: dict[str, Any] = {
            "$type": COLLECTIONS.algorithm,
            "name": name,
            "creator": self._require_did(),
        }
        if description:
            record["description"] = description
        if metadata is not None:
            record["metadata"] = json.dumps(metadata)
        return self._create_record(COLLECTIONS.algorithm, record)

    def create_project(
        self,
        *,
        name: str,
        phases: list[dict[str, Any]],
        description: Optional[str] = None,
    ) -> CreateRecordResult:
        record = {
            "$type": COLLECTIONS.project,
            "name": name,
            "creator": self._user_agent_self(),
            "phases": phases,
            **({"description": description} if description else {}),
        }
        return self._create_record(COLLECTIONS.project, record)

    # ---------------- queries ----------------

    def list_records(
        self,
        *,
        collection: str,
        repo: Optional[str] = None,
        limit: int = 50,
        cursor: Optional[str] = None,
    ) -> tuple[Optional[str], list[dict[str, Any]]]:
        repo_did = repo or self._require_did()
        resp = self.client.com.atproto.repo.list_records(
            params={"repo": repo_did, "collection": collection, "limit": limit, "cursor": cursor},
        )
        records = [
            {"uri": r.uri, "cid": r.cid, "value": dict(r.value) if hasattr(r.value, "__iter__") else r.value}
            for r in resp.records
        ]
        return resp.cursor, records

    def get_record(self, uri: str) -> dict[str, Any]:
        parsed = parse_at_uri(uri)
        resp = self.client.com.atproto.repo.get_record(
            params={"repo": parsed.repo, "collection": parsed.collection, "rkey": parsed.rkey},
        )
        return {"uri": resp.uri, "cid": getattr(resp, "cid", "") or "", "value": resp.value}

    def list_projects(self, *, repo: Optional[str] = None, limit: int = 50, cursor: Optional[str] = None):
        return self.list_records(collection=COLLECTIONS.project, repo=repo, limit=limit, cursor=cursor)

    def list_statements(self, *, repo: Optional[str] = None, limit: int = 50, cursor: Optional[str] = None):
        return self.list_records(collection=COLLECTIONS.statement, repo=repo, limit=limit, cursor=cursor)

    def list_votes(self, *, repo: Optional[str] = None, limit: int = 50, cursor: Optional[str] = None):
        return self.list_records(collection=COLLECTIONS.vote, repo=repo, limit=limit, cursor=cursor)
